### PR TITLE
[c10d FR analyzer] Output a meaningful debug report for users

### PR DESCRIPTION
**Pull request by fduwjj**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #134528

- This PR generates a more useful output log for users: P1552399180.
- It also fixes the logic when we check the all-gather size mismatch.
- Add dtype check for collective input/output
- We store more context information for error match_state so that we can report them in the file.
- Disable the size match for alltoall because we don't log the size for all inputs/outputs.
- Correct some types for func args specification.

cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @wz337 @wconstab @d4l3k @c-p-i-o

### CODE REVIEW

File: tools/flight_recorder/components/builder.py [hunk 1]
```diff
@@ -284,11 +300,17 @@ def build_collectives(
                     check_result, input_numel, output_numel = check_size_alltoall(
                         alltoall_cases
                     )
+                    # We don't log the input/output sizes for alltoall so we don't consider the size mismatch as an error for now.
+                    check_result = True
                     if not check_result:
```
Comment (c-p-i-o) [inline tools/flight_recorder/components/builder.py:305]: This `if` feels backwards?
I would have expected you to do something like:
```
check_result = False
if check_result:
```

File: tools/flight_recorder/components/builder.py [hunk 2]
```diff
@@ -284,11 +300,17 @@ def build_collectives(
...
+                        # When we see errors in all_to_all, it's hard to tell which rank is the source of the error.
                         mismatch[pg_name] += 1
                         print(
-                            f"Input/output mismatch in the collective for group {pg_name}:{desc} collective {profiling_name}",
-                            f"input_numel {input_numel} output_numel{output_numel}",
+                            f"Input/output mismatch in the collective {record_id} ",
+                            f"for group {pg_name}:{desc} collective {profiling_name} ",
+                            f"input_numel {input_numel} output_numel {output_numel} ",
+                            f"{input_sizes} {output_sizes} {len(expected_ranks)} {collective_state} ",
+                            f"\nCollective stack traces: \n{colletive_frames}",
```
Comment (c-p-i-o) [inline tools/flight_recorder/components/builder.py:313]: Typo here = unless it needs to be fixed elsewhere?
```suggestion
                            f"\nCollective stack traces: \n{collective_frames}",
```

File: tools/flight_recorder/components/builder.py [hunk 3]
```diff
@@ -306,11 +328,16 @@ def build_collectives(
...
+                error_msg = ", ".join(
+                    f"Error rank {error[0]}, {str(error[1])}" for error in errors
+                )
                 print(
-                    f"Collective errors for group {pg_name}:{desc} collective {profiling_name}",
-                    f"Found errors: {error_msg}",
+                    f"Collective {record_id} errors for group {pg_name}:{desc} collective {profiling_name} ",
+                    f"{input_sizes} {output_sizes} {len(expected_ranks)} {collective_state} ",
+                    f"\nFound errors: {error_msg}\n",
+                    f"\nCollective stack traces: \n{colletive_frames} ",
```
Comment (c-p-i-o) [inline tools/flight_recorder/components/builder.py:340]: Same as above. typo `colletive` -> `collective`

File: tools/flight_recorder/components/utils.py [hunk 4]
```diff
@@ -216,16 +230,46 @@ def check_version(versions: Dict[str, Any]) -> None:
...
+        entries (Dict[str, List[Dict[str, Any]]]): Entries sorted by record ID and filtered by the maximum starting point.
+    """
+
+    maximum_starting_record_id = 0
     for rank in entries:
+        # Since this is a ring buffer, we want to first sort the entries by `record_id` to
+        # find the largest starting point. For example, if the buffer has the following entries:
+        # Rank 0: [0, 1, 2, 3, 4, 5, 6]
+        # Rank 1: [7, 1, 2, 3, 4, 5, 6]
+        # Rank 2: [7, 8, 2, 3, 4, 5, 6]
```
Comment (c-p-i-o) [inline tools/flight_recorder/components/utils.py:260]: One thing to note is the expectation is entry `0` is the lowest record id and entry `N` is the largest record id. FR entries are supposed to be sorted when dumped.
So, we should not see 7, 8, 1, 2, 3 .

**Review by c-p-i-o — APPROVED** [review]
Minor comments - but LGTM!

File: tools/flight_recorder/components/builder.py (hunk 1 above)
Comment (fduwjj) [reply tools/flight_recorder/components/builder.py:305]: maybe I can rename it to `fail_check`

File: tools/flight_recorder/components/utils.py (hunk 4 above)
Comment (fduwjj) [reply tools/flight_recorder/components/utils.py:260]: Oh right, I missed that part, inside `dump_entries` we already have sorting already, then we don't need to re-sort it again.

**Pull request closed without merging**